BUG: DataFrame.where with category dtype

#### Code Sample (it is copy-pastable)

```python
import pandas as pd, numpy as np
df = pd.DataFrame(np.arange(2*3).reshape(2,3), columns=list('abc'))
mask = np.random.rand(*df.shape) < 0.5
df.where(mask)
# Output is correct:
#      a   b    c
# 0  NaN NaN  2.0
# 1  3.0 NaN  NaN

df.a = df.a.astype('category')
df.b = df.b.astype('category')
df.c = df.c.astype('category')
df.where(mask)
# ValueError: Wrong number of items passed 2, placement implies 1
# Expected output: the same as before, but now with dtype `category`.

df.a.where(mask[:,0])
# 0    NaN
# 1    3.0
# Name: a, dtype: float64
# should stay in dtype category

df.a.where(mask[:,0], other=None)
# 0    None
# 1    3
# Name: a, dtype: object
# Expected output: should stay in dtype category
```
#### Problem description

`df.where` should work with all dtypes, the [documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.where.html) doesn't say it works only for some dtypes. Also, NaNs are already correctly handled as missing data in `pd.Series` of type 'category', so one should be able to assign NaNs to them. Same with converting the dtype.

While writing this report I found that doing it column-by-column works correctly, so I'll use that as a workaround.

#### Output of ``pd.show_versions()``

<details>
# Paste the output here pd.show_versions() here

INSTALLED VERSIONS                                                                                                                                                                       [1/1839]
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-81-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.2
pytest: None
pip: 9.0.1
setuptools: 36.0.1
Cython: None
numpy: 1.13.1
scipy: 0.19.0
xarray: None
IPython: 6.1.0
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

Ubuntu  `lsb_release -a`:
------------
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.2 LTS
Release:        16.04
Codename:       xenial
</details>

Thanks for the guidance, @gfyoung!

While trying to add a test for the `.astype` error, I encountered another issue:

```
df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
mask = np.array([[True, False, True], [False, True, True]])

result = df.where(mask)
expected = DataFrame([[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"))

tm.assert_frame_equal(result, expected)
```

After running the code above, on some operating systems (`linux -- Python 3.6.7`, `win32 -- Python 3.6.7`, `win32 -- Python 3.7.5`) applying the boolean mask changes the dtype of the last column `[2, 5]` from `int64` to `int32`. That's the message from the logs:

```
>       tm.assert_frame_equal(result, expected)
E       AssertionError: Attributes of DataFrame.iloc[:, 2] are different
E       
E       Attribute "dtype" are different
E       [left]:  int32
E       [right]: int64
```

I cannot reproduce this behaviour locally.

`Output of pd.show_versions()`

<details>
INSTALLED VERSIONS ------------------ commit : 35e91f9 python : 3.6.5.final.0 python-bits : 64 OS : Darwin OS-release : 17.7.0 machine : x86_64 processor : i386 byteorder : little LC_ALL : None LANG : en_GB.UTF-8 LOCALE : en_GB.UTF-8
pandas : 0.26.0.dev0+734.g0de99558b.dirty
numpy : 1.17.3
pytz : 2019.3
dateutil : 2.8.0
pip : 19.3.1
setuptools : 39.0.1
Cython : 0.29.14
pytest : 5.2.2
hypothesis : 4.42.6
sphinx : 2.2.1
blosc : 1.8.1
feather : None
xlsxwriter : 1.2.2
lxml.etree : 4.4.1
html5lib : 1.0.1
pymysql : None
psycopg2 : None
jinja2 : 2.10.3
IPython : 7.9.0
pandas_datareader: None
bs4 : 4.7.1
bottleneck : 1.2.1
fastparquet : 0.3.2
gcsfs : None
lxml.etree : 4.4.1
matplotlib : 3.1.1
numexpr : 2.7.0
odfpy : None
openpyxl : 3.0.0
pandas_gbq : None
pyarrow : 0.15.1
pytables : None
s3fs : 0.3.5
scipy : 1.3.1
sqlalchemy : 1.3.10
tables : 3.6.1
xarray : 0.14.0
xlrd : 1.2.0
xlwt : 1.3.0
xlsxwriter : 1.2.2
</details>

@gfyoung : After further investigation, I identified where the problem is coming from.

When initialising df with int data w/o nans, the default dtype for all columns on some platforms (`linux -- Python 3.6.7`, `win32 -- Python 3.6.7` and `win32 -- Python 3.7.5`) is `int32`:
```
>       result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
>       expected = DataFrame( np.arange(2 * 3).reshape(2, 3), columns=list("ABC"), dtype=np.int64)
>       tm.assert_frame_equal(result, expected)
E       AssertionError: Attributes of DataFrame.iloc[:, 0] are different
E
E       Attribute "dtype" are different
E       [left]:  int32
E       [right]: int64
```

However, if you initialise df with int data w/ nans the dtype for columns w/o nans (same platforms) is `int64`:
```
>       result = DataFrame([[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"))
>       expected = DataFrame([[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int32 )
>       tm.assert_frame_equal(result, expected)
E       AssertionError: Attributes of DataFrame.iloc[:, 2] are different
E       
E       Attribute "dtype" are different
E       [left]:  int64
E       [right]: int32
```

On the other platforms (`linux -- Python 3.6.1`, `linux -- Python 3.6.9`, `linux -- Python 3.7.5` and `darwin -- Python 3.6.9`), both df with int data w/o nans and w/ nans are initialised with `int64` so there's no problem.

I believe this behaviour is unrelated to this issue as it's not testing the category preservation when using `.where()`. Furthermore, all the examples in the discussion have missed this behaviour (the columns have been [int, nan]/[nan, int] or [nan, nan] but not [int, int] alongside a column containing nan). I plan to change the input data in #29498 to the examples given in this issue and open a new issue describing the encountered behaviour. Do you agree with that?

--- a/pandas/tests/frame/test_dtypes.py
+++ b/pandas/tests/frame/test_dtypes.py
@@ -815,6 +815,16 @@ class TestDataFrameDataTypes:
         expected = concat([a1.astype(dtype), a2.astype(dtype)], axis=1)
         tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.parametrize(
+        "dtype", [{100: "float64", 200: "uint64"}, "category", "float64"]
+    )
+    def test_astype_column_metadata(self, dtype):
+        # GH 19920
+        columns = pd.UInt64Index([100, 200, 300], name="foo")
+        df = DataFrame(np.arange(15).reshape(5, 3), columns=columns)
+        df = df.astype(dtype)
+        tm.assert_index_equal(df.columns, columns)
+
     def test_df_where_change_dtype(self):
         # GH 16979
         df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
@@ -839,16 +849,6 @@ class TestDataFrameDataTypes:
         expected = DataFrame({"A": A, "B": B, "C": C})
 
         tm.assert_frame_equal(result, expected)
-
-    @pytest.mark.parametrize(
-        "dtype", [{100: "float64", 200: "uint64"}, "category", "float64"]
-    )
-    def test_astype_column_metadata(self, dtype):
-        # GH 19920
-        columns = pd.UInt64Index([100, 200, 300], name="foo")
-        df = DataFrame(np.arange(15).reshape(5, 3), columns=columns)
-        df = df.astype(dtype)
-        tm.assert_index_equal(df.columns, columns)
 
     @pytest.mark.parametrize("dtype", ["M8", "m8"])
     @pytest.mark.parametrize("unit", ["ns", "us", "ms", "s", "h", "m", "D"])

--- a/pandas/tests/frame/test_dtypes.py
+++ b/pandas/tests/frame/test_dtypes.py
@@ -1222,7 +1222,7 @@ class TestDataFrameDatetimeWithTZ:
         tm.assert_frame_equal(result, expected)
 
     def test_temp_64_nan(self):
-        result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        result = DataFrame([[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"))
         expected = DataFrame(
             [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int64
         )

--- a/pandas/tests/frame/test_dtypes.py
+++ b/pandas/tests/frame/test_dtypes.py
@@ -1214,6 +1214,20 @@ class TestDataFrameDatetimeWithTZ:
         )
         tm.assert_frame_equal(result, expected)
 
+    def test_temp_32_nan(self):
+        result = DataFrame([[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"))
+        expected = DataFrame(
+            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int32
+        )
+        tm.assert_frame_equal(result, expected)
+
+    def test_temp_64_nan(self):
+        result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        expected = DataFrame(
+            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int64
+        )
+        tm.assert_frame_equal(result, expected)
+
     def test_temp_32_mask(self):
         df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
         mask = np.array([[True, False, True], [False, True, True]])

--- a/pandas/tests/frame/test_dtypes.py
+++ b/pandas/tests/frame/test_dtypes.py
@@ -815,6 +815,42 @@ class TestDataFrameDataTypes:
         expected = concat([a1.astype(dtype), a2.astype(dtype)], axis=1)
         tm.assert_frame_equal(result, expected)
 
+    def test_temp_32(self):
+        result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        expected = DataFrame(
+            np.arange(2 * 3).reshape(2, 3), columns=list("ABC"), dtype=np.int32
+        )
+        tm.assert_frame_equal(result, expected)
+
+    def test_temp_64(self):
+        result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        expected = DataFrame(
+            np.arange(2 * 3).reshape(2, 3), columns=list("ABC"), dtype=np.int64
+        )
+        tm.assert_frame_equal(result, expected)
+
+    def test_temp_32_mask(self):
+        df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        mask = np.array([[True, False, True], [False, True, True]])
+
+        result = df.where(mask)
+        expected = DataFrame(
+            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int32
+        )
+
+        tm.assert_frame_equal(result, expected)
+
+    def test_temp_64_mask(self):
+        df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        mask = np.array([[True, False, True], [False, True, True]])
+
+        result = df.where(mask)
+        expected = DataFrame(
+            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int64
+        )
+
+        tm.assert_frame_equal(result, expected)
+
     @pytest.mark.parametrize(
         "dtype", [{100: "float64", 200: "uint64"}, "category", "float64"]
     )

--- a/pandas/tests/frame/test_dtypes.py
+++ b/pandas/tests/frame/test_dtypes.py
@@ -815,42 +815,6 @@ class TestDataFrameDataTypes:
         expected = concat([a1.astype(dtype), a2.astype(dtype)], axis=1)
         tm.assert_frame_equal(result, expected)
 
-    def test_temp_32(self):
-        result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
-        expected = DataFrame(
-            np.arange(2 * 3).reshape(2, 3), columns=list("ABC"), dtype=np.int32
-        )
-        tm.assert_frame_equal(result, expected)
-
-    def test_temp_64(self):
-        result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
-        expected = DataFrame(
-            np.arange(2 * 3).reshape(2, 3), columns=list("ABC"), dtype=np.int64
-        )
-        tm.assert_frame_equal(result, expected)
-
-    def test_temp_32_mask(self):
-        df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
-        mask = np.array([[True, False, True], [False, True, True]])
-
-        result = df.where(mask)
-        expected = DataFrame(
-            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int32
-        )
-
-        tm.assert_frame_equal(result, expected)
-
-    def test_temp_64_mask(self):
-        df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
-        mask = np.array([[True, False, True], [False, True, True]])
-
-        result = df.where(mask)
-        expected = DataFrame(
-            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int64
-        )
-
-        tm.assert_frame_equal(result, expected)
-
     @pytest.mark.parametrize(
         "dtype", [{100: "float64", 200: "uint64"}, "category", "float64"]
     )
@@ -1235,3 +1199,39 @@ class TestDataFrameDatetimeWithTZ:
             assert (
                 "2 2013-01-03 2013-01-03 00:00:00-05:00 2013-01-03 00:00:00+01:00"
             ) in result
+
+    def test_temp_32(self):
+        result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        expected = DataFrame(
+            np.arange(2 * 3).reshape(2, 3), columns=list("ABC"), dtype=np.int32
+        )
+        tm.assert_frame_equal(result, expected)
+
+    def test_temp_64(self):
+        result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        expected = DataFrame(
+            np.arange(2 * 3).reshape(2, 3), columns=list("ABC"), dtype=np.int64
+        )
+        tm.assert_frame_equal(result, expected)
+
+    def test_temp_32_mask(self):
+        df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        mask = np.array([[True, False, True], [False, True, True]])
+
+        result = df.where(mask)
+        expected = DataFrame(
+            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int32
+        )
+
+        tm.assert_frame_equal(result, expected)
+
+    def test_temp_64_mask(self):
+        df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        mask = np.array([[True, False, True], [False, True, True]])
+
+        result = df.where(mask)
+        expected = DataFrame(
+            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int64
+        )
+
+        tm.assert_frame_equal(result, expected)

--- a/pandas/tests/frame/test_dtypes.py
+++ b/pandas/tests/frame/test_dtypes.py
@@ -815,6 +815,31 @@ class TestDataFrameDataTypes:
         expected = concat([a1.astype(dtype), a2.astype(dtype)], axis=1)
         tm.assert_frame_equal(result, expected)
 
+    def test_df_where_change_dtype(self):
+        # GH 16979
+        df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        mask = np.array([[True, False, False], [False, False, True]])
+
+        result = df.where(mask)
+        expected = DataFrame(
+            [[0, np.nan, np.nan], [np.nan, np.nan, 5]], columns=list("ABC")
+        )
+
+        tm.assert_frame_equal(result, expected)
+
+        # change type to category
+        df.A = df.A.astype("category")
+        df.B = df.B.astype("category")
+        df.C = df.C.astype("category")
+
+        result = df.where(mask)
+        A = pd.Categorical([0, np.nan], categories=[0, 3])
+        B = pd.Categorical([np.nan, np.nan], categories=[1, 4])
+        C = pd.Categorical([np.nan, 5], categories=[2, 5])
+        expected = DataFrame({"A": A, "B": B, "C": C})
+
+        tm.assert_frame_equal(result, expected)
+
     @pytest.mark.parametrize(
         "dtype", [{100: "float64", 200: "uint64"}, "category", "float64"]
     )
@@ -1199,53 +1224,3 @@ class TestDataFrameDatetimeWithTZ:
             assert (
                 "2 2013-01-03 2013-01-03 00:00:00-05:00 2013-01-03 00:00:00+01:00"
             ) in result
-
-    def test_temp_32(self):
-        result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
-        expected = DataFrame(
-            np.arange(2 * 3).reshape(2, 3), columns=list("ABC"), dtype=np.int32
-        )
-        tm.assert_frame_equal(result, expected)
-
-    def test_temp_64(self):
-        result = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
-        expected = DataFrame(
-            np.arange(2 * 3).reshape(2, 3), columns=list("ABC"), dtype=np.int64
-        )
-        tm.assert_frame_equal(result, expected)
-
-    def test_temp_32_nan(self):
-        result = DataFrame([[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"))
-        expected = DataFrame(
-            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int32
-        )
-        tm.assert_frame_equal(result, expected)
-
-    def test_temp_64_nan(self):
-        result = DataFrame([[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"))
-        expected = DataFrame(
-            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int64
-        )
-        tm.assert_frame_equal(result, expected)
-
-    def test_temp_32_mask(self):
-        df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
-        mask = np.array([[True, False, True], [False, True, True]])
-
-        result = df.where(mask)
-        expected = DataFrame(
-            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int32
-        )
-
-        tm.assert_frame_equal(result, expected)
-
-    def test_temp_64_mask(self):
-        df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
-        mask = np.array([[True, False, True], [False, True, True]])
-
-        result = df.where(mask)
-        expected = DataFrame(
-            [[0, np.nan, 2], [np.nan, 4, 5]], columns=list("ABC"), dtype=np.int64
-        )
-
-        tm.assert_frame_equal(result, expected)

